HW0/P5/P5.txt

```diff
@@ -1 +1,10 @@
-# Your answers here.
+1. 255 seconds.
+2. Yes. There are 8 people totally. They can distribute 32 bags to one person and each person only count their own bags. 31+3=34 seconds. In this case, they just need 34 seconds to do it.
+3. If there are infinite number of people. They can use divide and conquer algorithm. They can use 128 people to sum each 2 bags first. And then sum them two by two at the same time and continue again until we have one sum. This strategy takes log(256)=8 seconds.
+4. If we have N bags. The fastest time to count them would be ceil(log2(N))
+5. If it takes 1 second to communicate between two cashier, it takes one more second to sum two number from two cashier. However, not every step takes 1 more second. At the first step of summing two bags, they don't need to communicate. Therefore, it takes log2(256)*2-1=15 seconds.
+6.
+   1 bag: It takes 1 second to give all bags to one cashier and this cashier takes 8 seconds to distribute 1 bag to everyone. After that, it takes 1 second to comunicate between every two cashier and 1 seconds to sum. And then continue this step until summing up to 1 number. As a result, it takes 1+8+8*2 = 25 seconds.
+   2 bags: It is similar to 1 bag but save 1 seconds in distributing bags and 1 seconds to sum up bags from 1 bag in two cashiers. There fore, it takes 25-1-1 = 23 seconds. 
+   There is a better strategy. First, I should give all bags to a cashier and this takes 1 second. And then this cashier can give 128 bags to another cashier and this takes another second. Every time the cashier do this, it takes 1 second. Now there is a problem: when to stop giving bags and count bags themselves. The answer is when they have 4 bags, they should count them themselves. If cashier give it to other, it takes 1 second to give it to another cashier, 1 second to sum two bags, 1 second to communicate and 1 second to sum two number, so it would be 4 seconds. However, if cashier counts himself, it just takes 3 seconds. 
+To sum up, they should give half of the bags in their hand to another cashier until they have 4 bags in hand and this strategy takes 7+3+6*2=22 seconds.
```

Comment:
I agree with all your answers except 6 where I found it hard to follow your argument, especially for 1bag, 2 bags